### DISCUSSION TITLE
Provide count distinct functionality

### DISCUSSION THREAD

**Discussion by TheCaffinatedDeveloper** [gorules/zen#266] [Ideas]
We have been loving the ZEN rules engine and it's ease of use. The javascript function "escape hatch" is quite nice and allows for some extensive capabilities. I will say however, the ability to count distinct elements in an array would be much appreciated. 

My recommendations: if the element in the array is a "zen primitive" such as number, date, string etc it should be fairly straightforward (and perhaps that's the first support). If its a _context_ then the user could provide a callback that returns a string of concatenated properties that make this context unique etc. 

I know this can be done using javascript functions but it would be nice to be able to do `count_distinct([date, date, date])` out of the box to calculate number of distinct dates for example. Or `len(distinct([date, date, date]))`

**Comment by ivanmiletic** [comment]
Good suggestion, added to backlog

**Comment by hatbrox** [comment]
use len + indexOf

My customer has made 10 product purchases and I want to count how many distinct product were bought on week end only (I added week end for extra fun).

**source data:**
```
{
  "customer": {
    "purchases": [
      { "name": "Product A", "purchaseDate": "2025-06-30" }, // Monday
      { "name": "Product B", "purchaseDate": "2025-07-01" }, // Tuesday
      { "name": "Product C", "purchaseDate": "2025-07-05" }, // Saturday
      { "name": "Product D", "purchaseDate": "2025-07-06" }, // Sunday
      { "name": "Product E", "purchaseDate": "2025-07-03" }, // Thursday
      { "name": "Product C", "purchaseDate": "2025-07-06" }, // Sunday
      { "name": "Product G", "purchaseDate": "2025-06-29" }, // Sunday
      { "name": "Product H", "purchaseDate": "2025-06-28" }, // Saturday
      { "name": "Product I", "purchaseDate": "2025-07-02" }, // Wednesday
      { "name": "Product J", "purchaseDate": "2025-07-04" }  // Friday
    ]
  }
}
```

**zen expression:**
```
len(
  filter(
    map(
      filter(customer.purchases, # => dayOfWeek(date(#.purchaseDate)) in [6, 7]),
      #.name
    ),
    (x, i, arr) => indexOf(arr, x) == i
  )
)

```
**result:**
Qty of "Product C" => 2 
Qty of all other products (D, G and H) => 1 each.

in total 5 products were purchased on the week end
**but only 4 distinct products were purchased on the week end**

**Comment by bknoles** [comment]
Did the API change? I'm getting an error on a local build of the editor when supplying a callback form like `(x, i, arr) => indexOf(arr, x) == i`, and the docs only show a `# > 1` type of callback.
